src/AppInstallerCLICore/Workflows/PortableInstallFlow.cpp

```diff
@@ -0,0 +1,309 @@
+            }
+            if (appsAndFeaturesEntry.DisplayVersion.empty())
+            {
+                appsAndFeaturesEntry.DisplayVersion = displayVersion;
+            }
+            if (appsAndFeaturesEntry.Publisher.empty())
+            {
+                appsAndFeaturesEntry.Publisher = publisher;
+            }
+            if (appsAndFeaturesEntry.ProductCode.empty())
+            {
+                appsAndFeaturesEntry.ProductCode = packageId;
+            }
+
+            return appsAndFeaturesEntry;
+        }
+
+        void AppendExeExtension(std::string& value)
+        {
+            if (!HasSuffix(value, ".exe"))
+            {
+                value += ".exe";
+            }
+        }
+
+        bool AddPortableLinksDirToPathRegistry(Manifest::ScopeEnum& scope, Utility::Architecture& arch, bool& isPathModified)
+        {
+            Key key;
+            if (scope == Manifest::ScopeEnum::Machine)
+            {
+                key = Registry::Key::CreateKeyAndOpen(HKEY_LOCAL_MACHINE, Normalize(s_PathSubkey_Machine));
+            }
+            else
+            {
+                key = Registry::Key::CreateKeyAndOpen(HKEY_CURRENT_USER, Normalize(s_PathSubkey_User));
+            }
+
+            std::string portableLinksDir = GetPortableLinksLocation(scope, arch).u8string();
+            AICLI_LOG(CLI, Info, << "Adding to Path environment variable: " << portableLinksDir);
+            std::wstring pathKey = { L"Path" };
+            std::string pathValue = key[pathKey]->GetValue<Value::Type::String>();
+            if (pathValue.find(portableLinksDir) == std::string::npos)
+            {
+                std::string modifiedPathValue = pathValue;
+
+                if (modifiedPathValue.back() != ';')
+                {
+                    modifiedPathValue += ";";
+                }
+
+                modifiedPathValue += portableLinksDir + ";";
+                bool result = key.SetKeyValue(pathKey, ConvertToUTF16(modifiedPathValue), REG_EXPAND_SZ);
+                if (result)
+                {
+                    isPathModified = true;
+                }
+            }
+
+            return true;
+        }
+
+        bool AddPortableEntryToUninstallRegistry(Manifest::ScopeEnum& scope, AppsAndFeaturesEntry& entry)
+        {
+            HKEY root = (scope == Manifest::ScopeEnum::Machine) ? HKEY_LOCAL_MACHINE : HKEY_CURRENT_USER;
+            std::wstring productCode = ConvertToUTF16(entry.ProductCode);
+            std::wstring displayName = ConvertToUTF16(entry.DisplayName);
+            std::wstring displayVersion = ConvertToUTF16(entry.DisplayVersion);
+            std::wstring publisher = ConvertToUTF16(entry.Publisher);
+            std::wstring uninstallString = L"winget uninstall --id " + productCode;
+            std::wstring fullRegistryKey = Normalize(s_UninstallSubkey) + L"\\" + productCode;
+
+            Key key = Key::CreateKeyAndOpen(root, fullRegistryKey);

```

Comment:
We *do not want* to be opening an existing key here unless it is an upgrade or a re-install over the existing key.  I think what that means is using the aforementioned `WinGetPackageIdentifier` and `WinGetSourceIdentifier` to ensure that this entry is for the same package.

Reply:
I added a couple checks to address this. The flow is as follows.

1. Call OpenIfExist() to determine if the entry in the registry already existed previously. If it didn't, we know we are installing a new portable package.
2. If this is not a new package, then I check the registry values to make sure that the `packageId` matches `WinGetPackageIdentifier` and `sourceId` matches `WinGetSourceIdentifier`. If both of those checks are true, then we can safely write to registry.

Lastly, if the user provides the `--portable-force` argument, then it bypasses these checks altogether.